Migration to html5 and bootstrap4 (#1681)
DjangoGirls/tutorial: en/django_forms/README.md

## Link to a page with the form

It's time to open `blog/templates/blog/base.html` in the code editor. In the `div` named `container` inside `page-header` `div` tag, we will add a link:

{% filename %}blog/templates/blog/base.html{% endfilename %}
```html
<a href="{% url 'post_new' %}" class="top-menu">

>     <svg width="1em" height="1em" viewBox="0 0 16 16" class="bi bi-file-earmark-plus" fill="currentColor" xmlns="http://www.w3.org/2000/svg"><path d="M4 0h5.5v1H4a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h8a1 1 0 0 0 1-1V4.5h1V14a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V2a2 2 0 0 1 2-2z"/><path d="M9.5 3V0L14 4.5h-3A1.5 1.5 0 0 1 9.5 3z"/><path fill-rule="evenodd" d="M8 6.5a.5.5 0 0 1 .5.5v1.5H10a.5.5 0 0 1 0 1H8.5V11a.5.5 0 0 1-1 0V9.5H6a.5.5 0 0 1 0-1h1.5V7a.5.5 0 0 1 .5-.5z"/></svg>

Reviewer: From a developer perspective I understand this, but if I was a student this would confuse me completely. I think this deserves a bit more explanation.

Author: I have thought about that. As it conserned to bootstrap4, they dont provide a icon package as css. This is where complexity begun. Next best option is font awesome. It is very similar to what it already have(glyphicon). But they have change the way of providing CND link publicly. I can still get the link from other CDN provides.

Then I thought of pushing the default icons of Bootstrap as SVG initially and change it to font awesome if required with an aditional commit to this PR.

If we are doing that, we need to fix up the best possible version of Font Awesome to be use.

One more thing is about using `integrity` and `crossorigin` attributes in `<link>` tags. @ekohl As you said, this decreases the readablility of code and increase complexity for users.

If we were to keep using the Bootstrap4 icons instead of Font awesome, I really feel like saving these svg files as assets and using it from `{% static 'img/icon-new-post.svg' %}` with `<img>` tag or `{% include './icons/new-psot.svg' %}`. The descriptive information regarding the same have to be added as well.

Author: @ekohl Please suggest how to proceed?

Reviewer: > One more thing is about using `integrity` and `crossorigin` attributes in `<link>` tags. @ekohl As you said, this decreases the readablility of code and increase complexity for users.

I don't mind these too much. On https://getbootstrap.com/ they use the same thing so it'd be the same thing if they were to follow that.

> If we were to keep using the Bootstrap4 icons instead of Font awesome, I really feel like saving these svg files as assets and using it from `{% static 'img/icon-new-post.svg' %}` with `<img>` tag or `{% include './icons/new-psot.svg' %}`. The descriptive information regarding the same have to be added as well.

This sounds like a nice solution.

## Link to a page with the form

It's time to open `blog/templates/blog/base.html` in the code editor. In the `div` named `container` inside `page-header` `div` tag, we will add a link:

{% filename %}blog/templates/blog/base.html{% endfilename %}
```html
<a href="{% url 'post_new' %}" class="top-menu">
    <svg width="1em" height="1em" viewBox="0 0 16 16" class="bi bi-file-earmark-plus" fill="currentColor" xmlns="http://www.w3.org/2000/svg"><path d="M4 0h5.5v1H4a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h8a1 1 0 0 0 1-1V4.5h1V14a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V2a2 2 0 0 1 2-2z"/><path d="M9.5 3V0L14 4.5h-3A1.5 1.5 0 0 1 9.5 3z"/><path fill-rule="evenodd" d="M8 6.5a.5.5 0 0 1 .5.5v1.5H10a.5.5 0 0 1 0 1H8.5V11a.5.5 0 0 1-1 0V9.5H6a.5.5 0 0 1 0-1h1.5V7a.5.5 0 0 1 .5-.5z"/></svg>
</a>
```

Note that we want to call our new view `post_new`. The [svg icon](https://icons.getbootstrap.com/icons/file-earmark-plus/) is provided by the [Bootstrap Icons](https://icons.getbootstrap.com/) and it will display a page icon with plus sign.

The addition of SVG code for the icon made it look very complicated. So we can try a different approach to achieve the same result.

> Let's save the svg code in a separate file. Save the following code in the file `blog/templates/blog/icons/add-post.svg`

Reviewer 2: I'd rather not make the tutorial longer than it already is. I'd also rather not introduce yet another language (We already have at least 4: command line (Bash or CMD), Python, HTML and CSS.) and thus would prefer to treat SVG just as an image format, without showing the SVG file's source code.

So can we let participants
- use the SVG as file without first having it inline in the template's HTML?
- download the whole SVG file (e.g. from https://raw.githubusercontent.com/twbs/icons/main/icons/file-earmark-plus.svg) and save it at the right location, rather than copying or typing its textual content?

For interested participants, we can link to https://en.wikipedia.org/wiki/Scalable_Vector_Graphics or some other information about the format. That'd be less overwhelming than the wall of ~~text~~ code we'd place before them with the changes currently suggested here.

Author: Will do that

Author: Updated https://github.com/DjangoGirls/tutorial/pull/1681/commits/56d846c0746237e1cf3f2758d5cf3d5d67cf4bce

Author: Please check the updated changes in the following sections:

https://django.nikz.in/en/django_forms/#link-to-a-page-with-the-form

https://django.nikz.in/en/django_forms/#edit-form

## Link to a page with the form

Before we add the link, we need some icons to use as buttons for the link. For this tutorial, download [file-earmark-plus.svg](https://raw.githubusercontent.com/twbs/icons/main/icons/file-earmark-plus.svg) and save it in the folder `blog/templates/blog/icons/`

> > Note: To download the svg image, click on the link and save the file using the keyboard shortcut `Ctrl + S`(Windows/Linux) or `Cmd + S`(MacOS).

Reviewer 2: Another (maybe better?) option: Open the context menu on the link (usually by right-click on it) and choose "Save as ..."

Author: @das-g I was thinking same thing. But then I remember users use different browsers and they have different content menu. For example, firefox show "Save link as". So if we use context menu, do we have to cover all browsers's context menu option to save?

Author: @das-g I tested the save from link option on the latest stable distributions of Firefox, Chrome, Edge, Safari, and IE 11. Found "Save link as" for all the browsers except Internet Explorer 11. Internet Explorer 11 has "Save target as" instead.

So should I change the note as follows?

> Note: To download the SVG image,  open the context menu on the link (usually by right-clicking on it) and choose "Save link as" option. Now you can navigate to the desired directory in the window on your screen and save the file.

Reviewer 2: Yes, that seems like a more natural way to save the file.

Funny that IE's menu entry label is actually more accurate, as it _is_ the link _target_ that's being saved (the file the link points at), and indeed not the link itself. :upside_down_face: (Off course, our description should use the more common "Save link as")

Author: Resolved in https://github.com/DjangoGirls/tutorial/pull/1681/commits/27483eb7bf7703eb6c02e51aaaf9702ceae631b8

> > Note: To download the svg image, click on the link and save the file using the keyboard shortcut `Ctrl + S`(Windows/Linux) or `Cmd + S`(MacOS).

Reviewer 2: Please write the abbreviation with capital letters:
> SVG

where it occurs in prose. (Don't change it in file names, as they are case sensitive on Linux.)

Author: Resolved in https://github.com/DjangoGirls/tutorial/pull/1681/commits/4d6ffbc4920caa307ba7a8e13619ce8b8237f4fc

> > Note: To download the SVG image, open the context menu on the link (usually by right-clicking on it) and choose "Save link as" option. Now you can navigate to the desired directory in the window on your screen and save the file.

Reviewer 2: ```suggestion
> Note: To download the SVG image, open the context menu on the link (usually by right-clicking on it) and select "Save link as". In the dialog asking you where to save the file, navigate to the `djangogirls` directory of your Django project, and within that to subdirectory `blog/templates/blog/icons/`, and save the file there.
```

> It's time to open `blog/templates/blog/base.html` in the code editor. Now we can use this icon file inside the base template as follow. In the `div` named `container` inside `page-header` `div` tag, we will add a link:

Reviewer 2: ```suggestion
It's time to open `blog/templates/blog/base.html` in the code editor. Now we can use this icon file inside the base template as follows. In the `div` named `container` inside `page-header` `div` tag, we will add a link:
```

Is "In the `div` named `container` inside `page-header` `div` tag" comprehensible to someone who's just begun learning HTML?

Author: @das-g True. I only followed the pattern. Similar pattern of referencing html tags and class through out the tutorial in different parts.

So how can we rewrite the sentence in a simpler terms?

Reviewer 2: > So how can we rewrite the sentence in a simpler terms?

We'd probably have to split it up into several sentences, maybe showing also the code before. E.g.

> It's time to open `blog/templates/blog/base.html` in the code editor. Now we can use this icon file inside the base template as follows. At the beginning of the `body` element (so, right after the tag `<body>`) you should have code like this:
> ```html
> <div class="page-header">
>     <div class="container">
>         <h1><a href="/">Django Girls Blog</a></h1>
>     </div>
> </div>
> ```
> Inside the inner `div` element (so, right after `<div class="container">`), we will add a link, just before the `<h1>...` part:

However, the explanation could remain vastly simpler, if there weren't nested `div` elements here. Are they really needed for this migration or would a single `div` element with two classes work, too? (See other comment in this review.)

{% filename %}blog/templates/blog/base.html{% endfilename %}
```html
<a href="{% url 'post_new' %}" class="top-menu">
    {% include './icons/file-earmark-plus.svg' %}
</a>
```

> Note that we want to call our new view `post_new`. The [SVG icon](https://icons.getbootstrap.com/icons/file-earmark-plus/) is provided by the [Bootstrap Icons](https://icons.getbootstrap.com/) and it will display a page icon with plus sign. We use a django template directive called `include`. This will inject the file's content in to django template since web browser know how to handle this type of content without any further processing.

Reviewer: ```suggestion
Note that we want to call our new view `post_new`. The [SVG icon](https://icons.getbootstrap.com/icons/file-earmark-plus/) is provided by the [Bootstrap Icons](https://icons.getbootstrap.com/) and it will display a page icon with plus sign. We use a Django template directive called `include`. This will inject the file's content into the Django template. The web browser knows how to handle this type of content without any further processing.
```

> > You can download the whole bootstrap icons [here](https://github.com/twbs/icons/releases/download/v1.1.0/bootstrap-icons-1.1.0.zip). Unzip the file and copy all the SVG image files in to a new folder inside `blog/templates/blog/` called `icons`. So that you can access a icon, e.g. pencil-fill.svg using the file path `blog/templates/blog/icons/pencil-fill.svg`

Reviewer: ```suggestion
> You can download the all Bootstrap icons [here](https://github.com/twbs/icons/releases/download/v1.1.0/bootstrap-icons-1.1.0.zip). Unzip the file and copy all the SVG image files into a new folder inside `blog/templates/blog/` called `icons`. So that you can access a icon, e.g. `pencil-fill.svg` using the file path `blog/templates/blog/icons/pencil-fill.svg`
```

Author: @ekohl 

"all the bootstrap.." is better than "the all bootstrap.." right??

Reviewer: Yes, that's what I meant.

> > You can download the whole bootstrap icons [here](https://github.com/twbs/icons/releases/download/v1.1.0/bootstrap-icons-1.1.0.zip). Unzip the file and copy all the SVG image files in to a new folder inside `blog/templates/blog/` called `icons`. So that you can access a icon, e.g. pencil-fill.svg using the file path `blog/templates/blog/icons/pencil-fill.svg`

Author: ```suggestion
> You can download all the Bootstrap icons [here](https://github.com/twbs/icons/releases/download/v1.1.0/bootstrap-icons-1.1.0.zip). Unzip the file and copy all the SVG image files into a new folder inside `blog/templates/blog/` called `icons`. So that you can access a icon, e.g. `pencil-fill.svg` using the file path `blog/templates/blog/icons/pencil-fill.svg`
```

Author: @ekohl Check this.

Reviewer 2: Is the "So that you can access a icon, e.g. `pencil-fill.svg` using the file path `blog/templates/blog/icons/pencil-fill.svg`" part necessary? Is so, it should either be part of the previous sentence ("so that" introduces a subclause)
```suggestion
> You can download all the Bootstrap icons [here](https://github.com/twbs/icons/releases/download/v1.1.0/bootstrap-icons-1.1.0.zip). Unzip the file and copy all the SVG image files into a new folder inside `blog/templates/blog/` called `icons`, so that you can access an icon like `pencil-fill.svg` using the file path `blog/templates/blog/icons/pencil-fill.svg`
```
or be rephrased to actually be a sentence of its own
```suggestion
> You can download all the Bootstrap icons [here](https://github.com/twbs/icons/releases/download/v1.1.0/bootstrap-icons-1.1.0.zip). Unzip the file and copy all the SVG image files into a new folder inside `blog/templates/blog/` called `icons`. That way you can access an icon like `pencil-fill.svg` using the file path `blog/templates/blog/icons/pencil-fill.svg`
```

Now we know how to add a new post. But what if we want to edit an existing one? This is very similar to what we just did. Let's create some important things quickly. (If you don't understand something, you should ask your coach or look at the previous chapters, since we covered all these steps already.)

> First, let's save the icon which represent edit button. Download [pencil-fill.svg](https://raw.githubusercontent.com/twbs/icons/main/icons/pencil-fill.svg) and save it in location `blog/templates/blog/icons/`

Reviewer: ```suggestion
First, let's save the icon which represent edit button. Download [pencil-fill.svg](https://raw.githubusercontent.com/twbs/icons/main/icons/pencil-fill.svg) and save it to the location `blog/templates/blog/icons/`.
```

> First, let's save the icon which represent edit button. Download [pencil-fill.svg](https://raw.githubusercontent.com/twbs/icons/main/icons/pencil-fill.svg) and save it to the location `blog/templates/blog/icons/`.

Reviewer 2: ```suggestion
First, let's save the icon which represents the edit button. Download [pencil-fill.svg](https://raw.githubusercontent.com/twbs/icons/main/icons/pencil-fill.svg) and save it to the location `blog/templates/blog/icons/`.
```
